File: src/api/routes/stats.ts
import { Context, router } from '../router.js';
import { getDbConnection } from '../../database/connection.js';

router.get('/api/stats', async ({ res }: Context) => {
  const db = getDbConnection();
  const total_users = (await db.get<{ count: number }>('SELECT COUNT(*) as count FROM users'))?.count ?? 0;
  const total_guilds = (await db.get<{ count: number }>('SELECT COUNT(*) as count FROM guilds'))?.count ?? 0;
  const total_messages = (await db.get<{ total: number }>('SELECT COALESCE(SUM(guild_messages), 0) as total FROM user_guilds'))?.total ?? 0;
  const avg_level = (await db.get<{ avg: number }>('SELECT COALESCE(AVG(overall_level), 0) as avg FROM users'))?.avg ?? 0;
  router.json(res, 200, { total_users, total_guilds, total_messages, avg_level });
});

File: api/src/routes/stats.rs
use axum::{extract::State, response::Json};
use sqlx::SqlitePool;
use serde_json::{json, Value};
use crate::auth::AuthUser;
use crate::models::Stats;

pub async fn get_stats(
    _auth: AuthUser,
    State(pool): State<SqlitePool>,
) -> Result<Json<Stats>, Json<Value>> {
    let stats = sqlx::query_as::<_, Stats>(
        r#"
        SELECT
            (SELECT COUNT(*) FROM users) as total_users,
            (SELECT COUNT(*) FROM guilds) as total_guilds,
            (SELECT COALESCE(SUM(guild_messages), 0) FROM user_guilds) as total_messages,
            (SELECT COALESCE(AVG(overall_level), 0.0) FROM users) as avg_level
        "#,
    )
    .fetch_one(&pool)
    .await
    .map_err(|e| Json(json!({ "error": e.to_string() })))?;

    Ok(Json(stats))
}
